execution/executor/src/components/in_memory_state_calculator_v2.rs

```diff
@@ -0,0 +1,254 @@
+        Ok(Some(CORE_CODE_ADDRESS))
+    }
+}
+
+/// Helper class for calculating state changes after a block of transactions are executed.
+pub struct InMemoryStateCalculatorV2 {
+    ///// These don't change during the calculation.
+    // This makes sure all in-mem nodes seen while proofs were fetched stays in mem during the
+    // calculation
+    _frozen_base: FrozenSparseMerkleTree<StateValue>,
+    proof_reader: ProofReader,
+
+    // The states at base version.
+    sharded_state_cache: [DashMap<StateKey, Option<StateValue>>; 16],
+    next_version: Version,
+    usage: StateStorageUsage,
+
+    //// These changes whenever make_checkpoint() happens.
+    checkpoint: SparseMerkleTree<StateValue>,
+    checkpoint_version: Option<Version>,
+    // This doesn't need to be frozen since `_frozen_base` holds a ref to the oldest ancestor
+    // already, but frozen SMT is used here anyway to avoid exposing the `batch_update()` interface
+    // on the non-frozen SMT.
+    latest: FrozenSparseMerkleTree<StateValue>,
+}
+
+impl InMemoryStateCalculatorV2 {
+    pub fn new(base: &StateDelta, state_cache: StateCache) -> Self {
+        let StateCache {
+            frozen_base,
+            state_cache,
+            proofs,
+        } = state_cache;
+        // TODO(grao): Only a base tree and version are needed because the current should be the
+        // same as base.
+        let StateDelta {
+            base,
+            base_version,
+            current,
+            current_version,
+            updates_since_base,
+        } = base.clone();
+
+        assert!(base_version == current_version);
+        assert!(updates_since_base.is_empty());
+
+        let sharded_state_cache = arr![DashMap::new(); 16];
+
+        // TODO(grao): Move this logic to an earlier stage.
+        let _: Vec<_> = state_cache
+            .into_iter()
+            .map(|(k, v)| {
+                sharded_state_cache[k.get_shard_id() as usize].insert(k, v);
+            })
+            .collect();
+        Self {
+            _frozen_base: frozen_base,
+            proof_reader: ProofReader::new(proofs),
+
+            sharded_state_cache,
+            next_version: current_version.map_or(0, |v| v + 1),
+            usage: current.usage(),
+
+            checkpoint: base,
+            checkpoint_version: base_version,
+            latest: current.freeze(),
+        }
+    }
+
+    pub fn calculate_for_transaction_block(
+        mut self,
+        to_keep: &[(Transaction, ParsedTransactionOutput)],
+        new_epoch: bool,
+    ) -> Result<(
+        Vec<HashMap<StateKey, Option<StateValue>>>,
+        Vec<Option<HashValue>>,
+        StateDelta,
+        Option<EpochState>,
+    )> {
+        let num_txns = to_keep.len();
+
+        let state_updates_vec = Self::get_state_updates(to_keep);
+
+        for (txn, txn_output) in to_keep.iter().take(num_txns - 1) {
+            assert!(!Self::need_checkpoint(txn, txn_output));
```

Comment:
how do we guarantee txn requiring checkpoints are not passed down to here?

Reply:
This is our assumption for block.